### ISSUE TITLE
Error on Deploy Policies, Initiatives, Policy Assignments step

### ISSUE THREAD

**Issue by MiguelElGallo** [Azure/enterprise-azure-policy-as-code#58]
Hello,
This is REPO is great!  We have waiting for this for so long! Congrats and cheers! 

I'm trying to deploy a very basic policy, and I get this error:

```
Delete obsolete and replaced Assignments (0)
Delete obsolete and replaced Initiative definitions (0)
Delete replaced Policy definitions (0)
---------------------------------------------------------------------------------------------------
Create new and replaced (create) Policy definitions (1)
    "emailalerts" - "EPACemailalerts"
Loaded Module 'Az.Accounts'
Loaded Module 'Az.MSGraph'
Loaded Module 'Az.Accounts'
Loaded Module 'Az.Authorization'
New-AzPolicyDefinition: /home/vsts/work/1/s/Scripts/Deploy/Deploy-AzPoliciesInitiativesAssignmentsFromPlan.ps1:195
Line |
 195 |          $null = New-AzPolicyDefinition @policyDefinition
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unexpected character encountered while parsing value: T. Path
     | '', line 0, position 0.

##[error]PowerShell exited with code '1'.
```

Attached is my policy and assignment.
Can you please let me know what is missing?
Thanks,
Miguel

```json
{
    "nodeName": "/Loc/",
    "assignment": {
        "name": "epacsqlalert",
        "displayName": "epac sql alerts",
        "description": "epac sql alerts"
    },
    "definitionEntry": {
        "policyName": "emailalerts",
        "friendlyNameToDocumentIfGuid": ""
    },
    "parameters": {
        "securityAlertPolicyEmail": {
            "value": "frompipe@test.com"
        }
    },
    "scope": {
        "epac-dev": [
            "/subscriptions/xxxxxxxxxxxx"
        ]
    }
}
```

```json
{
    "nodeName": "/Loc/",
    "assignment": {
        "name": "epacsqlalert",
        "displayName": "epac sql alerts",
        "description": "epac sql alerts"
    },
    "definitionEntry": {
        "policyName": "emailalerts",
        "friendlyNameToDocumentIfGuid": ""
    },
    "parameters": {
        "securityAlertPolicyEmail": {
            "value": "frompipe@test.com"
        }
    },
    "scope": {
        "epac-dev": [
            "/subscriptions/xxxxxxxxxxxx"
        ]
    }
}
```

```json
{
    "name": "emailalerts",
    "properties": {
        "displayName": "EPACemailalerts",
        "description": "EPACemailalerts",
        "assigmentName": "EPACemailalerts"
    },
        "parameters": {
        "securityAlertPolicyEmail": {
            "type": "String",
            "metadata": {
                "description": "The email address to send alerts",
                "displayName": "Email Address"
            }
        }
    },
    "policyRule": {
        "if": {
            "allOf": [
                {
                    "field": "type",
                    "equals": "Microsoft.Sql/managedInstances"
                }
            ]
        },
        "then": {
            "effect": "DeployIfNotExists",
            "details": {
                "type": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies",
                "name": "default",
                "existenceCondition": {
                    "allOf": [
                        {
                            "field": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies/state",
                            "equals": "Enabled"
                        }
                    ]
                },
                "roleDefinitionIds": [
                    "/providers/Microsoft.Authorization/roleDefinitions/056cd41c-7e88-42e1-933e-88ba6a50c9c3"
                ],
                "deployment": {
                    "properties": {
                        "mode": "incremental",
                        "template": {
                            "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
                            "contentVersion": "1.0.0.0",
                            "parameters": {
                                "location": {
                                    "type": "string"
                                },
                                "sqlManagedInstanceName": {
                                    "type": "string"
                                },
                                "sqlManagedInstanceDataBaseName": {
                                    "type": "string"
                                },
                                "securityAlertPolicyEmail": {
                                    "type": "string"
                                }
                            },
                            "variables": {},
                            "resources": [
                                {
                                    "type": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies",
                                    "apiVersion": "2017-03-01-preview",
                                    "name": "[concat(parameters('sqlManagedInstanceName'), '/', parameters('sqlManagedInstanceDataBaseName'), '/default')]",
                                    "properties": {
                                        "state": "Enabled",
                                        "disabledAlerts": [
                                            ""
                                        ],
                                        "emailAddresses": [
                                            "[parameters('securityAlertPolicyEmail')]"
                                        ],
                                        "emailAccountAdmins": true,
                                        "storageEndpoint": null,
                                        "storageAccountAccessKey": "",
                                        "retentionDays": 0
                                    }
                                }
                            ],
                            "outputs": {}
                        },
                        "parameters": {
                            "location": {
                                "value": "[field('location')]"
                            },
                            "sqlManagedInstanceName": {
                                "value": "[first(split(field('fullname'),'/'))]"
                            },
                            "sqlManagedInstanceDataBaseName": {
                                "value": "[field('name')]"
                            },
                            "securityAlertPolicyEmail": {
                                "value": "[parameters('securityAlertPolicyEmail')]"
                            }
                        }
                    }
                }
            }
        }
    }    
}
```

**Comment by techlake** [comment]
Your policy definition is flawed in some way. You can test this independent of EPAC. I included the whole fixed Policy below.

- Your "properties" needs to spans the rest of the Policy instead of only displayName, and description. (this caused your error)
- "assigmentName" is not a legal field value -> remove
- "disabledAlerts" should be an empty list, not a list with one empty string as the only entry

**Sidebar GitHub markdown tip** -- enclose your JSON with a line containing ` ```json ` and end the code block with ` ``` `. This makes it readable in GitHub itself (like the code below).

```json
{
    "name": "emailalerts",
    "properties": {
        "displayName": "EPACemailalerts",
        "description": "EPACemailalerts",
        "parameters": {
            "securityAlertPolicyEmail": {
                "type": "String",
                "metadata": {
                    "description": "The email address to send alerts",
                    "displayName": "Email Address"
                }
            }
        },
        "policyRule": {
            "if": {
                "allOf": [
                    {
                        "field": "type",
                        "equals": "Microsoft.Sql/managedInstances"
                    }
                ]
            },
            "then": {
                "effect": "DeployIfNotExists",
                "details": {
                    "type": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies",
                    "name": "default",
                    "existenceCondition": {
                        "allOf": [
                            {
                                "field": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies/state",
                                "equals": "Enabled"
                            }
                        ]
                    },
                    "roleDefinitionIds": [
                        "/providers/Microsoft.Authorization/roleDefinitions/056cd41c-7e88-42e1-933e-88ba6a50c9c3"
                    ],
                    "deployment": {
                        "properties": {
                            "mode": "incremental",
                            "template": {
                                "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
                                "contentVersion": "1.0.0.0",
                                "parameters": {
                                    "location": {
                                        "type": "string"
                                    },
                                    "sqlManagedInstanceName": {
                                        "type": "string"
                                    },
                                    "sqlManagedInstanceDataBaseName": {
                                        "type": "string"
                                    },
                                    "securityAlertPolicyEmail": {
                                        "type": "string"
                                    }
                                },
                                "variables": {},
                                "resources": [
                                    {
                                        "type": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies",
                                        "apiVersion": "2017-03-01-preview",
                                        "name": "[concat(parameters('sqlManagedInstanceName'), '/', parameters('sqlManagedInstanceDataBaseName'), '/default')]",
                                        "properties": {
                                            "state": "Enabled",
                                            "disabledAlerts": [],
                                            "emailAddresses": [
                                                "[parameters('securityAlertPolicyEmail')]"
                                            ],
                                            "emailAccountAdmins": true,
                                            "storageEndpoint": null,
                                            "storageAccountAccessKey": "",
                                            "retentionDays": 0
                                        }
                                    }
                                ],
                                "outputs": {}
                            },
                            "parameters": {
                                "location": {
                                    "value": "[field('location')]"
                                },
                                "sqlManagedInstanceName": {
                                    "value": "[first(split(field('fullname'),'/'))]"
                                },
                                "sqlManagedInstanceDataBaseName": {
                                    "value": "[field('name')]"
                                },
                                "securityAlertPolicyEmail": {
                                    "value": "[parameters('securityAlertPolicyEmail')]"
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by MiguelElGallo** [comment]
Hello, 
Thanks a lot for fixing the file.
I still get the error:
'''
Create new and replaced (create) Policy definitions (1)
    "emailalerts" - "EPACemailalerts"
Loaded Module 'Az.Accounts'
Loaded Module 'Az.MSGraph'
Loaded Module 'Az.Accounts'
Loaded Module 'Az.Authorization'
New-AzPolicyDefinition: /home/vsts/work/1/s/Scripts/Deploy/Deploy-AzPoliciesInitiativesAssignmentsFromPlan.ps1:195
Line |
 195 |          $null = New-AzPolicyDefinition @policyDefinition
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unexpected character encountered while parsing value: T. Path
     | '', line 0, position 0.

##[error]PowerShell exited with code '1'.
'''


This is the artifact that got generated 
'''
{
  "updatedPolicyDefinitions": {},
  "deletedInitiativeDefinitions": {},
  "expiredExemptions": {},
  "createdOn": "2022-08-18 06:03:18Z",
  "deletedExemptions": {},
  "deletedAssignments": {},
  "noChanges": false,
  "replacedExemptions": {},
  "updatedExemptions": {},
  "addedRoleAssignments": {
    "/subscriptions/xxxxxxxxxxxxxxxx/providers/Microsoft.Authorization/policyAssignments/epacsqlalert": {
      "roles": [
        {
          "scope": "/subscriptions/xxxxxxxxxxxxxxxx",
          "roleDefinitionId": "/providers/Microsoft.Authorization/roleDefinitions/056cd41c-7e88-42e1-933e-88ba6a50c9c3",
          "roleDisplayName": "SQL Security Manager"
        }
      ],
      "identity": null,
      "DisplayName": "epac sql alerts"
    },
    "/subscriptions/xxxxxxx/providers/Microsoft.Authorization/policyAssignments/epacsqlalert": {
      "roles": [
        {
          "scope": "/subscriptions/xxxxxxx",
          "roleDefinitionId": "/providers/Microsoft.Authorization/roleDefinitions/056cd41c-7e88-42e1-933e-88ba6a50c9c3",
          "roleDisplayName": "SQL Security Manager"
        }
      ],
      "identity": null,
      "DisplayName": "epac sql alerts"
    }
  },
  "replacedAssignments": {},
  "removedRoleAssignments": {},
  "rootScopeId": "/subscriptions/xxxxxxxxxx",
  "updatedAssignments": {},
  "deletedPolicyDefinitions": {},
  "newInitiativeDefinitions": {},
  "updatedInitiativeDefinitions": {},
  "replacedPolicyDefinitions": {},
  "newPolicyDefinitions": {
    "emailalerts": {
      "Metadata": null,
      "Description": "EPACemailalerts",
      "Name": "emailalerts",
      "Mode": "All",
      "Policy": {
        "if": {
          "allOf": [
            {
              "field": "type",
              "equals": "Microsoft.Sql/managedInstances"
            }
          ]
        },
        "then": {
          "effect": "DeployIfNotExists",
          "details": {
            "type": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies",
            "name": "default",
            "existenceCondition": {
              "allOf": [
                {
                  "field": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies/state",
                  "equals": "Enabled"
                }
              ]
            },
            "roleDefinitionIds": [
              "/providers/Microsoft.Authorization/roleDefinitions/056cd41c-7e88-42e1-933e-88ba6a50c9c3"
            ],
            "deployment": {
              "properties": {
                "mode": "incremental",
                "template": {
                  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
                  "contentVersion": "1.0.0.0",
                  "parameters": {
                    "location": {
                      "type": "string"
                    },
                    "sqlManagedInstanceName": {
                      "type": "string"
                    },
                    "sqlManagedInstanceDataBaseName": {
                      "type": "string"
                    },
                    "securityAlertPolicyEmail": {
                      "type": "string"
                    }
                  },
                  "variables": {},
                  "resources": [
                    {
                      "type": "Microsoft.Sql/managedInstances/databases/securityAlertPolicies",
                      "apiVersion": "2017-03-01-preview",
                      "name": "[concat(parameters('sqlManagedInstanceName'), '/', parameters('sqlManagedInstanceDataBaseName'), '/default')]",
                      "properties": {
                        "state": "Enabled",
                        "disabledAlerts": [],
                        "emailAddresses": [
                          "[parameters('securityAlertPolicyEmail')]"
                        ],
                        "emailAccountAdmins": true,
                        "storageEndpoint": null,
                        "storageAccountAccessKey": "",
                        "retentionDays": 0
                      }
                    }
                  ],
                  "outputs": {}
                },
                "parameters": {
                  "location": {
                    "value": "[field('location')]"
                  },
                  "sqlManagedInstanceName": {
                    "value": "[first(split(field('fullname'),'/'))]"
                  },
                  "sqlManagedInstanceDataBaseName": {
                    "value": "[field('name')]"
                  },
                  "securityAlertPolicyEmail": {
                    "value": "[parameters('securityAlertPolicyEmail')]"
                  }
                }
              }
            }
          }
        }
      },
      "SubscriptionId": "xxxxxxxxxxxxxxxxxxx",
      "DisplayName": "EPACemailalerts",
      "Parameter": {
        "securityAlertPolicyEmail": {
          "type": "String",
          "metadata": {
            "description": "The email address to send alerts",
            "displayName": "Email Address"
          }
        }
      }
    }
  },
  "orphanedExemptions": {},
  "newAssignments": {
    "/subscriptions/xxxxxxxxxxxxxxxx/providers/Microsoft.Authorization/policyAssignments/epacsqlalert": {
      "Metadata": {
        "roles": [
          {
            "scope": "/subscriptions/xxxxxxxxxxxxxxxx",
            "roleDefinitionId": "/providers/Microsoft.Authorization/roleDefinitions/056cd41c-7e88-42e1-933e-88ba6a50c9c3",
            "roleDisplayName": "SQL Security Manager"
          }
        ]
      },
      "Description": "epac sql alerts",
      "Name": "epacsqlalert",
      "managedIdentityLocation": "northeurope",
      "identityRequired": true,
      "policyId": "/subscriptions/xxxxxxxx/providers/Microsoft.Authorization/policyDefinitions/emailalerts",
      "PolicyParameterObject": {
        "securityAlertPolicyEmail": "frompipe@test.com"
      },
      "DisplayName": "epac sql alerts",
      "Scope": "/subscriptions/xxxxxxxxxxxxxxxx",
      "Id": "/subscriptions/xxxxxxxxxxxxxxxx/providers/Microsoft.Authorization/policyAssignments/epacsqlalert",
      "EnforcementMode": "Default",
      "NotScope": [
        "/subscriptions/xxxxxxxxxxxxxxxx"
      ]
    },
    "/subscriptions/xxxxxx/providers/Microsoft.Authorization/policyAssignments/epacsqlalert": {
      "Metadata": {
        "roles": [
          {
            "scope": "/subscriptions/xxxxxx",
            "roleDefinitionId": "/providers/Microsoft.Authorization/roleDefinitions/056cd41c-7e88-42e1-933e-88ba6a50c9c3",
            "roleDisplayName": "SQL Security Manager"
          }
        ]
      },
      "Description": "epac sql alerts",
      "Name": "epacsqlalert",
      "managedIdentityLocation": "northeurope",
      "identityRequired": true,
      "policyId": "/subscriptions/xxxxxxxxx/providers/Microsoft.Authorization/policyDefinitions/emailalerts",
      "PolicyParameterObject": {
        "securityAlertPolicyEmail": {
          "value": "frompipe@test.com"
        }
      },
      "DisplayName": "epac sql alerts",
      "Scope": "/subscriptions/xxxxxxx",
      "Id": "/subscriptions/xxxxxxxx/providers/Microsoft.Authorization/policyAssignments/epacsqlalert",
      "EnforcementMode": "Default"
    }
  },
  "tenantID": null,
  "rootScope": {
    "SubscriptionId": "xxxxxx"
  },
  "replacedInitiativeDefinitions": {},
  "newExemptions": {}
}
'''

**Comment by techlake** [comment]
Nothing strikes me as wrong. 
Before going the support route:
Can you try to add meta data with a category, please? If that works it would be a bug in EPAC

**Comment by MiguelElGallo** [comment]
Hi,
Added the meta data, and now the policy deployed correctly.
Bug #1?  :).
Thanks,
Miguel

**Comment by techlake** [comment]
This is a bug. I'll slot it into the next release. Likely late next week.

**Comment by martyh888** [comment]
This bug also occurs when there are no parameters and is omitted from the json which does occur so can we handle for this or if not then add some better error messages on the Confirm-ParametersMatch call

Currently fix is to add the line into the definition being blank

```json
{
    "name": "Newly created GUID",
    "properties": {
        "displayName": "Policy Display Name",
        "policyType": "Custom",
        "mode": "All",
        "description": "Policy Description",
        "metadata": {
            "version": "1.0.0",
            "category": "Your Category"
        },
        "parameters": {},
        "policyRule": {
            "if": {
                "Insert Logic Here"
            },
            "then": {
                "effect": "Audit, Deny, Modify, etc.",
                "details": {
                    "roleDefinitionIds": [],
                    "operations": []
                }
            }
        }
    }
}
```

**Comment by techlake** [comment]
@martyh888 Does this happen on New definitions??

**Comment by martyh888** [comment]
> @martyh888 Does this happen on New definitions??

Yes I created a new policy definition without parameters block as was not in a current policy in portal and got the error.  I haven't tried on an existing definition

**Comment by techlake** [comment]
Solved in v6.0
